File: VerbumKit/Sources/Models/BibleRelationship.swift
/// Spec §22.2.
public enum RelationshipType: String, Codable, Sendable, CaseIterable {
    case appearsIn
    case participatesIn
    case occursAt
    case occursDuring
    case references
    case relatedToTheme
    case relatedTo
    case precedes
    case follows
    case fulfills
    case quotes
}

/// Spec §22.2. Directed edge from `sourceId` to `targetId`.
public struct BibleRelationship: Identifiable, Codable, Equatable, Hashable, Sendable {
    public let id: String
    public let sourceId: String
    public let targetId: String
    public let type: RelationshipType
    /// Editorial confidence when known (spec §58). `nil` means unrated.
    public let confidence: Double?
    /// Provenance (spec §33). Every relationship shown to users should carry at least one.
    public let sourceReferenceIds: [String]

    public init(
        id: String,
        sourceId: String,
        targetId: String,
        type: RelationshipType,
        confidence: Double?,
        sourceReferenceIds: [String]
    ) {
        self.id = id
        self.sourceId = sourceId
        self.targetId = targetId
        self.type = type
        self.confidence = confidence
        self.sourceReferenceIds = sourceReferenceIds
    }
}

File: android/core/models/src/main/kotlin/com/nexussoft/verbum/models/BibleRelationship.kt
package com.nexussoft.verbum.models

/** docs/PRODUCT.md §22.2. `wireValue` is the string contract shared with iOS and the backend. */
enum class RelationshipType(val wireValue: String) {
    APPEARS_IN("appearsIn"),
    PARTICIPATES_IN("participatesIn"),
    OCCURS_AT("occursAt"),
    OCCURS_DURING("occursDuring"),
    REFERENCES("references"),
    RELATED_TO_THEME("relatedToTheme"),
    RELATED_TO("relatedTo"),
    PRECEDES("precedes"),
    FOLLOWS("follows"),
    FULFILLS("fulfills"),
    QUOTES("quotes");

    companion object {
        fun fromWireValue(value: String): RelationshipType? = entries.firstOrNull { it.wireValue == value }
    }
}

/** docs/PRODUCT.md §22.2. Directed edge from [sourceId] to [targetId]. */
data class BibleRelationship(
    val id: String,
    val sourceId: String,
    val targetId: String,
    val type: RelationshipType,
    /** Editorial confidence when known (§58). `null` means unrated. */
    val confidence: Double?,
    /** Provenance (§33). Every relationship shown to users should carry at least one. */
    val sourceReferenceIds: List<String>,
)
